### PR TITLE
Add bundle build dockerfile generation

### PR DESCRIPTION
**Pull request by ameyakhare**
Resolves TSK-889

## Description
Adds a unique dockerfile for generating an image from a bundle build.

## Test plan
Works locally.

### CODE REVIEW

File: pkg/build/builder.go [hunk 1]
```diff
@@ -333,6 +341,15 @@ func BuildDockerfile(c DockerfileConfig) (string, error) {
 	}
 }
 
+func BuildBundleDockerfile(c BundleDockerfileConfig) (string, error) {
+	switch c.BuildType {
+	case NodeBuildType:
```
Comment (leeweisberger) [inline pkg/build/builder.go:346]: Going to add other types after?

File: pkg/build/node.go [hunk 2]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
 	p.workspaces = workspacesObject.Packages
 	return nil
 }
+
+// node creates a dockerfile for Node (typescript/javascript).
```
Comment (leeweisberger) [inline pkg/build/node.go:633]: Use the correct function name. What's the diff between this and the node function?

File: pkg/build/node.go [hunk 3]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+
+		if err := json.Unmarshal(buf, &pkg); err != nil {
+			return "", fmt.Errorf("node: parsing %s - %w", rootPackageJSON, err)
+		}
+	}
+
+	for i, a := range buildArgs {
+		buildArgs[i] = fmt.Sprintf("ARG %s", a)
+	}
+	argsCommand := strings.Join(buildArgs, "\n")
```
Comment (leeweisberger) [inline pkg/build/node.go:693]: Split this into a helper and reuse

File: pkg/build/node.go [hunk 4]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+	if ok {
+		// Depending on how the options were serialized, the runtime can be
+		// either a string or TaskRuntime; handle both.
+		switch v := runtime.(type) {
+		case string:
+			isWorkflow = v == string(TaskRuntimeWorkflow)
+		case TaskRuntime:
+			isWorkflow = v == TaskRuntimeWorkflow
+		default:
+		}
```
Comment (leeweisberger) [inline pkg/build/node.go:671]: Split into a helper and reuse

File: pkg/build/node.go [hunk 5]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+		NodeVersion:         GetNodeVersion(options),
+		PreInstallCommand:   pkg.Settings.PreInstallCommand,
+		PostInstallCommand:  pkg.Settings.PostInstallCommand,
+		Args:                argsCommand,
+		IsWorkflow:          isWorkflow,
+		InlineParserScript:  parser.NodeParserScript,
+		InlineUniversalShim: inlineString(universalNodeShim),
+	}
+
+	// TODO: install hooks per root directory
```
Comment (leeweisberger) [inline pkg/build/node.go:710]: What does this mean? You're doing install hooks below. What are we missinig?

File: pkg/build/node.go [hunk 6]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+			// image lean. This doesn't apply to Yarn v2 (specifically Plug'n'Play), which uses the
+			// cache directory for storing packages.
+			installCommand = "yarn install --non-interactive --production --frozen-lockfile && yarn cache clean"
+		}
+	} else if hasPackageLock {
+		// Use npm ci if possible, since it's faster and behaves better:
+		// https://docs.npmjs.com/cli/v8/commands/npm-ci
+		installCommand = "npm ci --production"
+	}
+	cfg.InstallCommand = strings.ReplaceAll(installCommand, "\n", "\\n")
```
Comment (leeweisberger) [inline pkg/build/node.go:812]: break this into a helper

File: pkg/build/node.go [hunk 7]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+			&& {{.InlineWorkflowBundlerScript}} >> /airplane/.airplane/workflow-bundler.js
+		RUN node /airplane/.airplane/workflow-bundler.js
+		{{end}}
+
+		RUN {{.InlineUniversalShim}} > /airplane/.airplane/universal-shim.js && \
+			esbuild /airplane/.airplane/universal-shim.js \
+				--bundle \
+				--platform=node {{.ExternalFlags}} \
+				--target=node{{.NodeVersion}} \
+				--outfile=/airplane/.airplane/dist/universal-shim.js
```
Comment (leeweisberger) [inline pkg/build/node.go:904]: You should just build this once and then just copy it into the image. We don't need to build this for every single task that is deployed.

File: pkg/build/node.go [hunk 8]
```diff
@@ -621,3 +629,287 @@ func (p *pkgJSONWorkspaces) UnmarshalJSON(data []byte) error {
...
+
+		RUN esbuild {{.BuildFilesList}} \
+			--bundle \
+			--platform=node {{.ExternalFlags}} \
+			--target=node{{.NodeVersion}} \
+			--outdir=/airplane/.airplane \
+			--outbase=/airplane
+
+		RUN node /airplane/.airplane/inlineParser.js {{.DiscoverFilesList}}
```
Comment (leeweisberger) [inline pkg/build/node.go:913]: How are we going to get this out of the image?

File: pkg/build/node.go [hunk 9]
```diff
@@ -43,6 +44,10 @@ type templateParams struct {
 	PreInstallPath                   string
 	PostInstallPath                  string
 	PackageCopyCmds                  []string
+	BuildFilesList                   string
+	DiscoverFilesList                string
```
Comment (leeweisberger) [inline pkg/build/node.go:48]: Can you add comments what these are? There's a loooota stuffs here

File: pkg/build/node.go (hunk 7 above)
Comment (leeweisberger) [reply pkg/build/node.go:904]: We can do as a followup, but there's no reason to do this every time. We can just add a build script next to the shim that produces the built js file.

**Review by leeweisberger — CHANGES_REQUESTED** [review]

File: pkg/build/builder.go (hunk 1 above)
Comment (ameyakhare) [reply pkg/build/builder.go:346]: yep, thats the plan

File: pkg/build/node.go (hunk 5 above)
Comment (ameyakhare) [reply pkg/build/node.go:710]: I fixed this. basically, we're only going to look for a pre/post path in the task root for bundle builds

File: pkg/build/node.go (hunk 8 above)
Comment (ameyakhare) [reply pkg/build/node.go]: follow up, not doing it right now

**Review by leeweisberger — CHANGES_REQUESTED** [review]
re-request after tests plz

File: pkg/build/node.go [hunk 10]
```diff
@@ -44,6 +45,13 @@ type templateParams struct {
 	PreInstallPath                   string
 	PostInstallPath                  string
 	PackageCopyCmds                  []string
+
+	// string of space-separated entity files to esbuild (user code)
+	FilesToBuild string
+	// string of space-separated built js files to discover entity configs from
```
Comment (leeweisberger) [inline pkg/build/node.go:51]:
```suggestion
	// FilesToDiscover represents the built JS files to discover entity configs from. These built files are the output of esbuild.
```

File: pkg/build/node.go [hunk 11]
```diff
@@ -654,3 +632,317 @@ func ReadPackageJSON(path string) (PackageJSON, error) {
...
+				--outfile=/airplane/.airplane/dist/universal-shim.js
+
+		RUN esbuild {{.FilesToBuild}} \
+			--bundle \
+			--platform=node {{.ExternalFlags}} \
+			--target=node{{.NodeVersion}} \
+			--outdir=/airplane/.airplane \
+			--outbase=/airplane
+
+		RUN node /airplane/.airplane-build-tools/inlineParser.js {{.FilesToDiscover}}
```
Comment (leeweisberger) [inline pkg/build/node.go:890]:
```suggestion
                # Discover inline tasks now that dependencies are installed and tasks are built.
                # FilesToDiscover should be the location where esbuild outputs transpiled JS files.
		RUN node /airplane/.airplane-build-tools/inlineParser.js {{.FilesToDiscover}}
```

**Review by leeweisberger — APPROVED** [review]

**Pull request merged**